### PR TITLE
Add support for piercing projectiles

### PR DESCRIPTION
**Pull request by Waselon**
## About The Pull Request

Railgun projectiles now can penetrate mobs/objects/turfs.
Adds support for 'piercing' type of ammo.

## Why It's Good For The Game

Bravemole wanted to see this on railgun first, so here it is. Also slightly buffs underused gun.

## Changelog
:cl:
balance: railgun projectiles now penetrate stuff
code: added support for piercing projectiles
/:cl:

### CODE REVIEW

**Comment by Lumipharon** [comment]
I am hyped for next level FF.

**Comment by Moggram** [comment]
> I am hyped for next level FF.

I can see it now. bro fires the railgun and hits 5 guys. adman plays [this](https://www.youtube.com/watch?v=cGsuGAb5NUQl)

**Review by BraveMole — CHANGES_REQUESTED** [review]
Plus what i said on disc

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/projectile.dm [hunk 1]
```diff
@@ -822,6 +833,11 @@ So if we are on the 32th absolute pixel coordinate we are on tile 1, but if we a
 //----------------------------------------------------------
 
 /atom/proc/bullet_act(obj/projectile/proj)
+	if(((proj.ammo.flags_ammo_behavior & AMMO_PASS_THROUGH_MOVABLE) || (proj.ammo.flags_ammo_behavior & AMMO_PASS_THROUGH_TURF)) && HAS_TRAIT(proj, TRAIT_PROJ_HIT_SOMETHING))
```
Comment (BraveMole) [inline code/modules/projectiles/projectile.dm:836]: (proj.ammo.flags_ammo_behavior & AMMO_PASS_THROUGH_MOVABLE|AMMO_PASS_THROUGH_TURF) works. But uh, is it even possible to call bullet_act twice if you are not AMMO_PASS_THROUGH anyway? In that case the flags check is useless

File: code/modules/projectiles/projectile.dm [hunk 2]
```diff
@@ -61,6 +61,8 @@
 	var/list/atom/movable/uncross_scheduled = list() // List of border movable atoms to check for when exiting a turf.
 
 	var/damage = 0
+	var/penetration = 0
+	var/sundering = 0
```
Comment (BraveMole) [inline code/modules/projectiles/projectile.dm:65]: New vars must be autodoced

/// A comment on what is penetration
var/penetration = 0

File: code/modules/projectiles/ammo_datums.dm [hunk 3]
```diff
@@ -38,6 +38,7 @@ GLOBAL_LIST_INIT(no_sticky_resin, typecacheof(list(/obj/item/clothing/mask/faceh
 	var/barricade_clear_distance	= 1			// How far the bullet can travel before incurring a chance of hitting barricades; normally 1.
 	var/armor_type					= "bullet"	// Does this have an override for the armor type the ammo should test? Bullet by default
 	var/sundering					= 0 		// How many stacks of sundering to apply to a mob on hit
+	var/on_pierce_multiplier 		= 1			// A multiplier applied to piercing projectile, that reduces its damage/penetration/sundering on hit
```
Comment (BraveMole) [inline code/modules/projectiles/ammo_datums.dm:41]: ditto, need autodoc

File: code/modules/projectiles/projectile.dm [hunk 4]
```diff
@@ -822,6 +833,11 @@ So if we are on the 32th absolute pixel coordinate we are on tile 1, but if we a
 //----------------------------------------------------------
 
 /atom/proc/bullet_act(obj/projectile/proj)
+	if(((proj.ammo.flags_ammo_behavior & AMMO_PASS_THROUGH_MOVABLE) || (proj.ammo.flags_ammo_behavior & AMMO_PASS_THROUGH_TURF)) && HAS_TRAIT(proj, TRAIT_PROJ_HIT_SOMETHING))
+		proj.damage *= proj.ammo.on_pierce_multiplier
+		proj.penetration *= proj.ammo.on_pierce_multiplier
+		proj.sundering *= proj.ammo.on_pierce_multiplier
+	ADD_TRAIT(proj, TRAIT_PROJ_HIT_SOMETHING, TRAIT_PROJ_HIT_SOMETHING)
```
Comment (BraveMole) [inline code/modules/projectiles/projectile.dm:840]: it should be ADD_TRAIT(proj, TRAIT_PROJ_HIT_SOMETHING, BULLET_ACT_TRAIT)

the last argument is the "origin" of the trait

**Review by BraveMole — CHANGES_REQUESTED** [review]

**Comment by BraveMole** [comment]
Not sure anyone used the railgun BUT this didn't fucked at all with regular bullet so good to go

**Pull request merged**